stdlib/asm/crypto/fri/frif2e2.masm

```diff
@@ -0,0 +1,279 @@
+    mem_loadw                               #[0,0,a1,a0,add'-2,t_d,e1,e0,p,d,poe,add'-2,..]
+    push.0.0                                #[0,0,0,0,a1,a0,add'-2,t_d,e1,e0,p,d,poe,add'-2,..]
+    movup.6                                 #[add'-2,0,0,0,0,a1,a0,t_d,e1,e0,p,d,poe,add'-2,..]
+    mem_loadw                               #[C,a1,a0,t_d,e1,e0,p,d,poe,add'-2,..]
+    swapw.2                                 #[e0,p,d,poe,a1,a0,t_d,e1,C,add'-2,..]
+    swap.2                                  #[d,p,e0,poe,a1,a0,t_d,e1,C,add'-2,..]
+    movup.7                                 #[e1,d,p,e0,poe,a1,a0,t_d,C,add'-2,..]
+    movup.7                                 #[t_d,e1,d,p,e0,poe,a1,a0,C,add'-2,..]
+    movupw.2                                #[C,t_d,e1,d,p,e0,poe,a1,a0,add'-2,..]
+    movup.7                                 #[p,C,t_d,e1,d,e0,poe,a1,a0,add'-2,..]
+    movup.7                                 #[d,p,C,t_d,e1,e0,poe,a1,a0,add'-2,..]
+end
+
+#! The following procedure computes an iteration of FRI verification for a query.
+#! The stack starts in the following initial configuration [d,p,C,t_d,e1,e0,poe,a1,a0,..] where d is 
+#! the current domain size, p the current position, C is the commitment of current layer, t_d 
+#! its corresponding tree depth, (e1,e0) is the evaluation point, poe is the position in the exponent
+#! of the domain generator, (a1,a0) is the alpha to compute the next evaluation point.
+#! TODO: Check where/if some checks are needed at the beginning, like p < d (p&d are u32) and that 2^t_d == d 
+#! Input: #[d,p,C,t_d,e1,e0,poe,a1,a0,add',add_f..]
+#! Output: #[t_d-1,e1,e0,p%d/2,d/2,poe_sq,add',add_f..]
+export.verify_query_layer
+    u32unchecked_div.2          #[d/2,p,C,t_d,...]
+    dup                         #[d/2,d/2,p,C,t_d,...]
+    dup.2 swap                  #[d/2,p,d/2,p,C,t_d,...]
+    u32checked_divmod           #[p%d/2,?,d/2,p,C,t_d,...]
+    swapw                       #[C,p%d/2,?,d/2,p,t_d,...]
+    dup.4                       #[p%d/2,C,p%d/2,?,d/2,p,t_d,...]
+    dup.9                       #[t_d,p%d/2,C,p%d/2,?,d/2,p,t_d,...]
+    mtree_get  swapw            #[C,V',p%d/2,?,d/2,p,t_d,...]  V' are the digest of values V = (v3,v2,v1,v0)
+    
+    dropw
+    adv.keyval                  # load the leaf values onto the adv tape
+                                #[V',p%d/2,?,d/2,p,t_d,e1,e0,poe,a1,a0,add',add_f,...]
+
+    dup.15                      #[add_f,V',p%d/2,?,d/2,p,t_d,e1,e0,poe,a1,a0,add',add_f,...]
+
+    push.4.0.0.0                #[0,0,0,4,add_f,V',p%d/2,?,d/2,p,t_d,...]
+    push.0.0.0.0.0.0.0.0        #[0,0,0,0,0,0,0,0,0,0,0,4,add_f,V',p%d/2,?,d/2,p,t_d,...]
+    adv_pipe                    #[T,D,S,add_f,V',p%d/2,?,d/2,p,t_d,...]
+    dropw swapw dropw           #[D,add_f,V',p%d/2,?,d/2,p,t_d,...]
+    movup.4  movdn.8            #[D,V',add_f,p%d/2,?,d/2,p,t_d,...]
+    eqw drop dropw movup.4      #[add_f,V',p%d/2,?,d/2,p,t_d,...]
+    sub.2 mem_loadw             #[V,p%d/2,?,d/2,p,t_d,...]
```

Comment:
This section is to get pre-image of the leaf node onto the stack, right? If I counted it correctly, this requires about 50 VM cycles.

Since the leaf contains fewer than 8 elements, one optimization could be to use `rpperm` instead of `adv_pipe`. I think this could save quite a few cycles.

Also, I see there is an `eqw drop` sequence on line 137. Wouldn't this just drop the result of the comparison?

Reply:
> Since the leaf contains fewer than 8 elements, one optimization could be to use rpperm instead of adv_pipe. I think this could save quite a few cycles.

I can try this approach and compare to the current one.

> Also, I see there is an eqw drop sequence on line 137. Wouldn't this just drop the result of the comparison?

This is a mistake. It should assert that the words are equal i.e. `drop` should be `assert`

Reply:
I think with the following sequence we should be able to do it in about 36 cycles:
```
push.4.0.0.0.0.0.0.0
adv_loadw
loc_storew.0
push.0.0.0.0
rpperm
swapw3
movup.4
assert_eq
movup.3
assert_eq
movup.2
assert_eq
assert_eq
dropw
loc_loadw.0
```

Reply:
Integrated your proposal.